posthog/api/search.py

```diff
@@ -0,0 +1,77 @@
+import re
+from typing import Any
+from django.contrib.postgres.search import SearchQuery, SearchRank, SearchVector
+from django.db.models import Model, Value, CharField, F, QuerySet
+from django.db.models.functions import Cast
+from django.http import HttpResponse
+from rest_framework import viewsets
+from rest_framework.permissions import IsAuthenticated
+from rest_framework.request import Request
+from rest_framework.response import Response
+
+from posthog.api.routing import StructuredViewSetMixin
+from posthog.permissions import ProjectMembershipNecessaryPermissions, TeamMemberAccessPermission
+from posthog.models import Action, Cohort, Insight, Dashboard, FeatureFlag, Experiment, Team
+
+
+class SearchViewSet(StructuredViewSetMixin, viewsets.ViewSet):
+    permission_classes = [IsAuthenticated, ProjectMembershipNecessaryPermissions, TeamMemberAccessPermission]
+
+    def list(self, request: Request, **kw) -> HttpResponse:
+        query = process_query(request.GET.get("q", "").strip())
+        counts = {}
+
+        # empty queryset to union things onto it
+        qs = Dashboard.objects.annotate(type=Value("empty", output_field=CharField())).filter(team=self.team).none()
+
+        for klass in (Action, Cohort, Insight, Dashboard, Experiment, FeatureFlag):
+            klass_qs, type = class_queryset(klass, team=self.team, query=query)
+            qs = qs.union(klass_qs)
+            counts[type] = klass_qs.count()
+
+        return Response({"results": qs, "counts": counts})
+
+
+UNSAFE_CHARACTERS = r"[\'&|!<>():]"
+"""Characters unsafe in a `tsquery`."""
+
+
+def process_query(query: str):
+    """
+    Converts a query string into a to_tsquery compatible string, where
+    the last word is a prefix match. This allows searching as you type.
+    """
+    query = re.sub(UNSAFE_CHARACTERS, " ", query).strip()
+    query = re.sub(r"\s+", " & ", query)  # combine words with &
+    if len(query) == 0:
+        return None
+    query += ":*"  # prefix match last word
```

Comment:
One annoying bit: if I search for "week", I won't get any results with "weekly". So perhaps every word should be prefix-matched? Not sure if this would be a better overall experience than if we ended up using trigrams.